### PR TITLE
fix_cursor: do not change line number when edit will not impact cursor row

### LINKED ISSUE #13691: nvim_buf_set_line: cursor loses line position if end of range is equal to or greater than the number of lines in the buffer
- `nvim --version`: lateset nightly
- `vim -u DEFAULTS` (version: ) behaves differently? no
- Operating system/version: macOS/Linux 
- Terminal name/version: iterm/Alacritty
- `$TERM`: xterm-256color

I'm working on a PR for this. The culprit is:

https://github.com/neovim/neovim/blob/adbf2dd86b8c7d898843bdf47b89b423f62bee4b/src/nvim/api/buffer.c#L1885-L1886

### Steps to reproduce using `nvim -u NORC`
1. `:lua vim.api.nvim_buf_set_lines(nil, 0, 3, false, { "line_1", "line_2", "line_3", "line_4"})`
2. Put your cursor on line 2 or 3
3. `:lua vim.api.nvim_buf_set_lines(nil, 0, 4, false, { "updated_line_1", "updated_line_2", "updated_line_3"})`

The cursor line position will be lost, this does not occur if the same action is repeated with the end range being before the last line of the buffer

1. `:lua vim.api.nvim_buf_set_lines(nil, 0, 3, false, { "line_1", "line_2", "line_3", "line_4"})`
2. Put your cursor on line 2 or 3
3. `:lua vim.api.nvim_buf_set_lines(nil, 0, 3, false, { "updated_line_1", "updated_line_2", "updated_line_3"})`

### Actual behaviour
The cursor line position will be lost on nvim_set_buf_lines if the end range is greater to or equal the number of buffer lines.

### Expected behaviour
* If the end range is equal to or greater than the number of buffer lines, the cursor line position should still be preserved.
* If the cursor will be on the last line of the buffer after the set_lines operation, the cursor position should be preserved.
* If the cursor position will be on an invalid line after the set_lines operation, I'm not sure if it would be better for the cursor to be set to the last line, or to the beginning of the buffer. I would guess the last line?

### PR DESCRIPTION
**Pull request by mjlbach**
Closes #13691, related: https://github.com/neovim/nvim-lspconfig/issues/465

Not sure if this is the right thing to do (not sure why extra was being used, but cursor was uniformly being set to the beginning row of edit range), but it fixes the issue. 

Edit: If this is the desired behavior, I'll write a test.

### CODE REVIEW

**Comment by mjlbach** [comment]
@bfredl I think you and @justinmk were the last people to review a change here.  #9961 readded this function after it was removed in #9674

**Comment by mjlbach** [comment]
@bfredl Added test

Result pre-patch:

```
[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional/lua/buffer_updates_spec.lua @ 228: lua buffer event callbacks: on_lines has valid cursor position while deleting lines
test/helpers.lua:73: Expected objects to be the same.
Passed in:
(number) 1
Expected:
(number) 2
```

Result post-patch:
```
[----------] Running tests from test/functional/lua/buffer_updates_spec.lua
[ RUN      ] lua buffer event callbacks: on_lines has valid cursor position while deleting lines: 1.94 ms OK
```

**Comment by mjlbach** [comment]
Three tests are failing that didn't fail before
```
2021-01-31T17:14:52.7100529Z Failures: 
2021-01-31T17:14:52.7100926Z 	From test_python3.vim:
2021-01-31T17:14:52.7101503Z 	Found errors in Test_Write_To_Current_Buffer_Fixes_Cursor_Clear():
2021-01-31T17:14:52.7106816Z 	Caught exception in Test_Write_To_Current_Buffer_Fixes_Cursor_Clear(): Vim(return):Error invoking 'python_execute' on channel 3 (python3-script-host):Traceback (most recent call last):  File "<string>", line 1, in <module>  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/buffer.py", line 65, in __setitem__    return self.request('nvim_buf_set_lines', start, end, False, lines)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/common.py", line 58, in request    return self._session.request(name, self, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/nvim.py", line 182, in request    res = self._session.request(name, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/msgpack_rpc/session.py", line 104, in request    raise self.error_wrapper(err)pynvim.api.common.NvimError: Vim:E315: ml_get: invalid lnum: 11 @ function RunTheTest[37]..Test_Write_To_Current_Buffer_Fixes_Cursor_Clear[3]..provider#python3#Call, line 18
2021-01-31T17:14:52.7110779Z 	Found errors in Test_Write_To_Current_Buffer_Fixes_Cursor_List():
2021-01-31T17:14:52.7115885Z 	Caught exception in Test_Write_To_Current_Buffer_Fixes_Cursor_List(): Vim(return):Error invoking 'python_execute' on channel 3 (python3-script-host):Traceback (most recent call last):  File "<string>", line 1, in <module>  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/buffer.py", line 65, in __setitem__    return self.request('nvim_buf_set_lines', start, end, False, lines)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/common.py", line 58, in request    return self._session.request(name, self, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/nvim.py", line 182, in request    res = self._session.request(name, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/msgpack_rpc/session.py", line 104, in request    raise self.error_wrapper(err)pynvim.api.common.NvimError: Vim:E315: ml_get: invalid lnum: 11 @ function RunTheTest[37]..Test_Write_To_Current_Buffer_Fixes_Cursor_List[3]..provider#python3#Call, line 18
2021-01-31T17:14:52.7119946Z 	Found errors in Test_Write_To_Current_Buffer_Fixes_Cursor_Str():
2021-01-31T17:14:52.7124938Z 	Caught exception in Test_Write_To_Current_Buffer_Fixes_Cursor_Str(): Vim(return):Error invoking 'python_execute' on channel 3 (python3-script-host):Traceback (most recent call last):  File "<string>", line 1, in <module>  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/buffer.py", line 61, in __setitem__    return self.request('nvim_buf_set_lines', i, i + 1, True, lines)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/common.py", line 58, in request    return self._session.request(name, self, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/api/nvim.py", line 182, in request    res = self._session.request(name, *args, **kwargs)  File "/home/runner/.local/lib/python3.8/site-packages/pynvim/msgpack_rpc/session.py", line 104, in request    raise self.error_wrapper(err)pynvim.api.common.NvimError: Vim:E315: ml_get: invalid lnum: 11 @ function RunTheTest[37]..Test_Write_To_Current_Buffer_Fixes_Cursor_Str[3]..provider#python3#Call, line 18
```

**Comment by mjlbach** [comment]
Ok, now some tests with the python provider are failing

**Comment by mjlbach** [comment]
All due to the setter on buffer in pynvim

Running:
```
py3 vim.current.buffer.request('nvim_buf_set_lines', 0, 11, False, [])
```
In the following buffer:
```

9
8
7
6
5
4
3
2
1
0
```
Leads to
```
Error invoking 'python_execute' on channel 6 (python3-script-host):
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/michael/.local/miniconda3/envs/neovim/lib/python3.6/site-packages/pynvim/api/common.py", line 58, in request
    return self._session.request(name, self, *args, **kwargs)
  File "/home/michael/.local/miniconda3/envs/neovim/lib/python3.6/site-packages/pynvim/api/nvim.py", line 182, in request
    res = self._session.request(name, *args, **kwargs)
  File "/home/michael/.local/miniconda3/envs/neovim/lib/python3.6/site-packages/pynvim/msgpack_rpc/session.py", line 104, in request
    raise self.error_wrapper(err)
pynvim.api.common.NvimError: Vim:E315: ml_get: invalid lnum: 11
```

File: test/functional/lua/buffer_updates_spec.lua [hunk 1]
```diff
@@ -225,6 +225,18 @@ describe('lua buffer event callbacks: on_lines', function()
     eq(1, meths.get_var('listener_cursor_line'))
   end)
 
+  it('has valid cursor position while deleting lines', function()
+   meths.buf_set_lines(0, 0, -1, true, { "line_1", "line_2", "line_3", "line_4"})
+    exec_lua([[
```
Comment (bfredl) [inline test/functional/lua/buffer_updates_spec.lua:230]: is there a difference versus just using `meths.win_set_cursor` ?

File: test/functional/lua/buffer_updates_spec.lua [hunk 2]
```diff
@@ -225,6 +225,18 @@ describe('lua buffer event callbacks: on_lines', function()
...
+      vim.api.nvim_win_set_cursor(nil, {2, 0})
+    ]])
+    local cursor_row = exec_lua([[ return vim.api.nvim_win_get_cursor(0)[1] ]])
```
Comment (bfredl) [inline test/functional/lua/buffer_updates_spec.lua:233]: ditto

File: test/functional/lua/buffer_updates_spec.lua [hunk 3]
```diff
@@ -225,6 +225,18 @@ describe('lua buffer event callbacks: on_lines', function()
...
+    eq(2, cursor_row)
+    meths.buf_set_lines(0, 0, -1, true, { "line_1", "line_2", "line_3"})
+    cursor_row = exec_lua([[ return vim.api.nvim_win_get_cursor(0)[1] ]])
```
Comment (bfredl) [inline test/functional/lua/buffer_updates_spec.lua:236]: ditto

**Comment by mjlbach** [comment]
I don't think the test failure on win32 is due to this pr?

```
[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional\core\channels_spec.lua @ 61: channels can use stdio channel
test\helpers.lua:73: Expected objects to be the same.
Passed in:
(table: 0x0ef2c338) {
  [1] = 'notification'
 *[2] = 'stdout'
  [3] = {
    [1] = 3
    [2] = {
      [1] = '' } } }
Expected:
(table: 0x0ef287d8) {
  [1] = 'notification'
 *[2] = 'stderr'
  [3] = {
    [1] = 3
    [2] = {
      [1] = '' } } }
```

**Comment by mjlbach** [comment]
Yeah, that test is just flakey, also see failure in this PR https://github.com/neovim/neovim/runs/1801487593

**Pull request merged**